src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3249,12 +3357,21 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
                 case ConversionKind.ExplicitDynamic:
                 case ConversionKind.ImplicitDynamic:
                     resultAnnotation = operandType.IsNull ? NullableAnnotation.Unknown : operandType.NullableAnnotation;
+                    if (resultAnnotation == NullableAnnotation.NotNullable && targetType.IsTypeParameter())
+                    {
+                        resultAnnotation = NullableAnnotation.NotNullableBasedOnAnalysis;
```

Comment:
>resultAnnotation = NullableAnnotation.NotNullableBasedOnAnalysis; [](start = 24, length = 65)

Is this necessary, or is this an optimization (caching)?

From our previous discussion, code that consumes types with annotations should be able to handle `NotNullable` (ie. it should check the `Type` to see if it's a type parameter). So I think that building the `resultType` with `targetType` and `NotNullable` is okay too. Is that correct?

Reply:
> So I think that building the resultType with targetType and NotNullable is okay too. Is that correct?

No it is not correct. In this case result is not nullable regardless of type, ```NotNullable``` would leave subject to interpretation based on type.

---
In reply to: [236465468](https://github.com/dotnet/roslyn/pull/31264#discussion_r236465468) [](ancestors = 236465468)

Reply:
I thought that all consumers of TSWA are expected to handle all the cases (ie. needs to check the type when encountering `NotNullable`): 
"Any time a decision has to be made whether something annotated by NotAnnotated can/can't be null, annotation and the type should be looked at together."
[https://github.com/dotnet/roslyn/pull/30913/files#r231669404](https://github.com/dotnet/roslyn/pull/30913/files#r231669404)

Did something change?

---
In reply to: [236470034](https://github.com/dotnet/roslyn/pull/31264#discussion_r236470034) [](ancestors = 236470034,236465468)

Reply:
>"Any time a decision has to be made whether something annotated by NotAnnotated can/can't be null, annotation and the type should be looked at together."
[https://github.com/dotnet/roslyn/pull/30913/files#r231669404](https://github.com/dotnet/roslyn/pull/30913/files#r231669404)

>Did something change?

No, this didn't change. ```NotAnnotated``` sometimes means - might be null depending on type. ```NotNullableBasedOnAnalysis``` always means - might not be null. They are not equivalent.

---
In reply to: [236471607](https://github.com/dotnet/roslyn/pull/31264#discussion_r236471607) [](ancestors = 236471607,236470034,236465468)

Reply:
Thanks for the clarification.

I think the difference hinges on looking at `targetType.IsTypeParameter()` rather than `targetType.IsPossiblyNullableReferenceTypeTypeParameter()`.
Why are we not looking at the latter? Here's a scenario to illustrate the question:

```C#
class Program
{
    public static void Main()
    {
        M<string>(null);
    }

    static void M<T>(dynamic d) where T : class
    {
        T local = (T)d;
        // null-state of local should be "maybe null"
        if (local is null)
        {
            Console.Write("Was null");
        }
    }
}
```

---
In reply to: [236472810](https://github.com/dotnet/roslyn/pull/31264#discussion_r236472810) [](ancestors = 236472810,236471607,236470034,236465468)

Reply:
> Why are we not looking at the latter? 

Because it is not necessary to look at it. The current logic is simpler and always produces correct result. 

> Here's a scenario to illustrate the question

We will not get into the ```if``` for this scenario.

---
In reply to: [236476862](https://github.com/dotnet/roslyn/pull/31264#discussion_r236476862) [](ancestors = 236476862,236472810,236471607,236470034,236465468)

Reply:
> We will not get into the if for this scenario.

I just ran that code and it printed "Was null".

---
In reply to: [236478000](https://github.com/dotnet/roslyn/pull/31264#discussion_r236478000) [](ancestors = 236478000,236476862,236472810,236471607,236470034,236465468)

Reply:
> I just ran that code and it printed "Was null".

I meant the ```If``` in the implementation ("for this scenario"), not the ```if``` "in the scenario".

---
In reply to: [236478317](https://github.com/dotnet/roslyn/pull/31264#discussion_r236478317) [](ancestors = 236478317,236478000,236476862,236472810,236471607,236470034,236465468)